### ISSUE TITLE
WOZ: Balance of Power

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#1022]
Matthew Eric D'Asaro said on Slack (21 Jan 2022):
>OK. Here is my weird woz of the wee-hours. This is the 1986 version of "Balance of Power". It will not boot at all in AppleWin, but works perfectly in MAME and OpenEmulator, the opposite of what I have usually found. But here is the kicker. It will boot perfectly consistently on real hardware (enhanced Apple IIe), but ONLY with one of my two identical and seemingly-working Apple 5.25" drives. The non-working drive passes all of the applesauce tests as far as I can tell and has consistently booted every other title I have tried with it. Very weird. I am assuming this is a random-bit-density issue?

Kent Dickey:
>This image found another KEGS bug: This WOZ uses 36 tracks, and there was code in KEGS forcing 35 tracks at most. This Woz is a little unusual--there's no track 34.75 or 35.25, just 35.0 (which is the 36th track).

[Balance of Power - Disk 1, Side A.zip](https://github.com/AppleWin/AppleWin/files/7908737/Balance.of.Power.-.Disk.1.Side.A.zip)

**Comment by tomcw** [comment]
(My post to Slack on 21 Jan)
>Looks like a track sync issue: start at T$00 (at a precise position), seek to T$23, read 256 nibbles and find a sequence of 12 nibbles in the 256 we've just read.

**Comment by medasaro** [comment]
I have confirmed that this is still a bug in 1.30.11.0. The "Balance of Power" .woz image works with MAME and with a BMOW Floppy Emu on real hardware, but does not work in AppleWin.

-Matthew D'Asaro

**Comment by tomcw** [comment]
Thanks Matthew - I'll take another look at this.

**Comment by tomcw** [comment]
Here's the cross-track sync protection check code:

Summary:
- Seek to T$00
- Read Address Field for S$0A
- Seek to T$35
- Read $100 nibbles to $DC00
- Search for the 13 byte sequence: F2,F3; F4,F5,F6,F7, F8,F9,FA,FB, FC,FD,FE

```
ff18
	lda #0		; phase 0 / trk 0
	jsr d133
	
	ldx $bf30	; #$60
ff20	jsr d398	; READ ADDRESS FIELD (Beneath Apple ProDOS 120 and 130 Supplement.pdf)
		bcs ff73	; fail
		lda d36e
		cmp #$0a	; sector = $0A ?
		bne ff20

ff2c
	lda #$46	; phase 70 / trk 35
	jsr d133	; ARM MOVE ROUTINE (Beneath Apple ProDOS 120 and 130 Supplement.pdf)
	(40) = Y = #0
	(41) = A = #$DC
	pha
ff3a
	r c0ee	; read mode
	r c0ec	; latch
	read $100 nibbles to $dc00
	
	lda #0
	jsr d133
	r c0e8	; motor off

	pla
	(41) = A	; #$DC
				; Y = 0
ff52	lda	($40),y
		iny
		beq ff73	; fail
		cmp #$f2
		bne ff52
		lda ($40),y
		cmp #$f3
		bne ff52
		; found f2,f3
		sta $ff		; #$f3
		ldx #0
ff65		inc $ff		; next target: #$f4, #$f5, ... #$fe
			iny
			lda ($40),y
			inx
			cpx #$0c
			beq ff91	; pass?
			cmp $ff
			beq ff65
ff73	;fail			
		r c0e8
		:

ff91
	rts
```

**Comment by tomcw** [comment]
The original .woz from @medasaro had `optimalBitTiming=31` (on Slack he said he was trying multiple bit timings to see if it make a difference), but this is wrong, and messes with the track head movement, so here's a new image with `optimalBitTiming=32`:

[Balance of Power(OptTiming0x20) - Disk 1, Side A.zip](https://github.com/AppleWin/AppleWin/files/9387077/Balance.of.Power.OptTiming0x20.-.Disk.1.Side.A.zip)

NB. The woz with `optimalBitTiming=31` _does_ work in MAME, but I don't know how it is implemented there, ie. it could be wrong, in that (eg) it uses optimalBitTiming=32 during track head movement.

Anyway, the correct value is `optimalBitTiming=32`.

**Comment by tomcw** [comment]
Here's the T$23 data, roughly where the head must be after seeking from T$00:

```
A600:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
A700:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
A800:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
A900:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
AA00:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
AB00:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF
AC00:   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FF   FC   FF(2)FF(2)FF
AD04:(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF
AE44:(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)FF(2)F2   F3   F4   F5   F6   F7   F8   F9   FA   FB   FC   FD   FE   FF   FF   FF   FF   FF
```

AppleWin, for optimalBitTiming=32, was starting too early at (bit offset) $a4d0.

NB. for optimalBitTiming=31 it was starting after this critical data sequence, at (bit offset) $b298.

**Comment by tomcw** [comment]
Details on the AppleWin fix:

The WOZ 2.1 spec says to use this formula:
```
new_position = current_position * new_track_length / current_track_length
```

Using integer maths, this will round down each time, slicing off a fraction of a nibble each time.
Since the code is moving across `35*2` track phases, then that's potentially `70*8` (0x230) bit cells.

So the fix was just to correct this by modifying the above formula to additionally add 1.

This turns out to be just good enough with the critical sequence ($F2,F3,...$FE) occurring in the $DC00 buffer at $DCF0-DCFC.

---

Some footnotes:

1) Modifying the formula to "add 2" also works, and the critical sequence occurs more centrally in the $DC00 buffer at $DCAA-DCB6.

But Glutton (#1021), the other cross-track sync sensitive game fails ~50% of the time.

2) Modifying the formula to use float maths works a bit better, but isn't quite good enough:

```
			double byte = ((double)currentPosition * (double)pFloppy->m_nibbles) / (double)currentTrackLength;
			pFloppy->m_byte = (int) ceil(byte);
```

The bit offset start position is $A5D8 and after reading the $100 nibbles, the bit offset is $AE10, which is still in the sync-FF field.

**Comment by tomcw** [comment]
The improvement in 5888a2b gives better fidelity to the "+1 nibble" fix above, as it now operates on the bit offset (instead of nibble offset).

So using a "+7 bit" fix adjustment means the critical sequence ($F2,F3,...$FE) occurs in the $DC00 buffer at $DCDD-DCE9.
(And Glutton still booting 100%)

NB. As expected the original woz with optimalBitTiming=31 still fails.

**Comment by tomcw** [comment]
4am attached another 'Balance of Power' woz to Slack on 22nd Aug (`applesauce` channel).
This has optimalBitTiming=32, and works in both AppleWin (with the above fix) and MAME 0.236.

```
AppleWin: (Using above bitOffset "+7 bit" fix)
Succeeds at FF91, with: Y=C6
DCB9: F2 F3 ...
DCC5: FE

MAME 0.236:
Succeeds at FF91, with: Y=CE
DCC1: F2 F3 ...
DCCD: FE
```

**Comment by tomcw** [comment]
@medasaro - fixed in this new [AppleWin 1.30.12.0](https://github.com/AppleWin/AppleWin/releases/tag/v1.30.12.0).
